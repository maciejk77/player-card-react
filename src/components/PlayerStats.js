import React from 'react';
import '../styles/player_stats.css'
import '../styles/badges.css';

const PlayerStats = (props) => {

  // set of player data
  const player = props.data[0].player
  const first_name = player.name.first
  const last_name = player.name.last
  const position = player.info.positionInfo.split(' ').pop()
  
  // fetching club badge with matching CSS class
  const team_name = (props.data[0].player.currentTeam.name.toLowerCase().split(' '));
  const club_badge_class = team_name.length > 1 ? `${team_name[0]}-${team_name[1]}` : `${team_name[0]}`;

  // remapping stats object to have key/value pair e.g. {name: goals, value: 5} ==> {goals: 5, ...}
  const remapped_data = props.data[0].stats.reduce((previousValue, currentValue) => {
    previousValue[currentValue.name] = currentValue.value;
    return previousValue;
  }, {})

  // set of player statistics
  const { appearances, goals, goal_assist, backward_pass, fwd_pass, mins_played } = remapped_data
  const goals_per_match = (goals / appearances).toFixed(2)
  const passes_per_minute = ((backward_pass + fwd_pass) / mins_played).toFixed(2)
  
  return (
    <div className="card">
      
      <div className="wrapper">
        <div className="card__player-image"><img src={require(`../assets/p${props.data[0].player.id}.png`)} style={{width: "20rem"}} alt="" /></div>
        <div className="card__club-badge"><div className={club_badge_class}></div></div> 
      </div>
     
      <div className="info">
        <div className="info__name">{first_name} {last_name}</div>
        <div className="info__position">{position}</div>
      </div>
      
      <div className="stats">  
        <div className="stats__item">Appearances<span className="stats__item-data">{appearances}</span></div>
        <div className="stats__item">Goals<span className="stats__item-data">{goals}</span></div>
        <div className="stats__item">Assists<span className="stats__item-data">{goal_assist ? goal_assist : 'No Data'}</span></div>
        <div className="stats__item">Goals per match<span className="stats__item-data">{goals_per_match}</span></div>
        <div className="stats__item">Passes per minute<span className="stats__item-data">{passes_per_minute}</span></div>
      </div>

    </div>
  )

}

export default PlayerStats;



