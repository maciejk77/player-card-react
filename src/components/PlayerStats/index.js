import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { fetchPlayerData } from '../../actions';
import '../../styles/player_stats.css';
import '../../styles/badges.css';

const PlayerStats = ({ data, fetchPlayerData }) => {
  useEffect(() => {
    fetchPlayerData();
  });

  // set of player data
  const { player, stats } = data[0];
  const { first_name, last_name } = player.name;
  const position = player.info.positionInfo.split(' ').pop();

  // fetching club badge with matching CSS class
  const team_name = player.currentTeam.name.toLowerCase().split(' ');
  const club_badge_class =
    team_name.length > 1
      ? `${team_name[0]}-${team_name[1]}`
      : `${team_name[0]}`;

  // remapping stats object to have key/value pair e.g. {name: goals, value: 5} ==> {goals: 5, ...}
  const remapped_data = stats.reduce((previousValue, currentValue) => {
    previousValue[currentValue.name] = currentValue.value;
    return previousValue;
  }, {});

  // set of player statistics
  const {
    appearances,
    goals,
    goal_assist,
    backward_pass,
    fwd_pass,
    mins_played
  } = remapped_data;
  const goals_per_match = (goals / appearances).toFixed(2);
  const passes_per_minute = ((backward_pass + fwd_pass) / mins_played).toFixed(
    2
  );

  return (
    <div className="card">
      <div className="wrapper">
        <div className="card__player-image">
          <img
            src={require(`../../assets/p${player.id}.png`)}
            style={{ width: '20rem' }}
            alt=""
          />
        </div>
        <div className="card__club-badge">
          <div className={club_badge_class}></div>
        </div>
      </div>

      <div className="info">
        <div className="info__name">
          {first_name} {last_name}
        </div>
        <div className="info__position">{position}</div>
      </div>

      <div className="stats">
        <div className="stats__item">
          Appearances<span className="stats__item-data">{appearances}</span>
        </div>
        <div className="stats__item">
          Goals<span className="stats__item-data">{goals}</span>
        </div>
        <div className="stats__item">
          Assists
          <span className="stats__item-data">
            {goal_assist ? goal_assist : 'No Data'}
          </span>
        </div>
        <div className="stats__item">
          Goals per match
          <span className="stats__item-data">{goals_per_match}</span>
        </div>
        <div className="stats__item">
          Passes per minute
          <span className="stats__item-data">{passes_per_minute}</span>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  return { player: state.player.players };
};

export default connect(
  mapStateToProps,
  { fetchPlayerData }
)(PlayerStats);
