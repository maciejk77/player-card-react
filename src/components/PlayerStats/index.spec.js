import React from 'react';
import { shallow } from 'enzyme';
import PlayerStats from './index';
import { players } from '../../data/data.json';

describe('<PlayerStats />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PlayerStats data={players} />);
  });

  it('renders one <PlayerStats /> component', () => {
    expect(wrapper.length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
});
