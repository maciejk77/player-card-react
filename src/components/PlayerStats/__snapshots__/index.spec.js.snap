// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerStats /> renders one <PlayerStats /> component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(PlayerStats)
    data={
        Array [
            Object {
              "player": Object {
                "age": "27 years 139 days",
                "currentTeam": Object {
                  "id": 21,
                  "name": "Tottenham Hotspur",
                  "shortName": "Spurs",
                  "teamType": "FIRST",
                },
                "id": 4916,
                "info": Object {
                  "position": "D",
                  "positionInfo": "Centre/Right Central Defender",
                  "shirtNum": 4,
                },
                "name": Object {
                  "first": "Toby",
                  "last": "Alderweireld",
                },
                "nationalTeam": Object {
                  "country": "Belgium",
                  "demonym": "Belgian",
                  "isoCode": "BE",
                },
              },
              "stats": Array [
                Object {
                  "name": "goals",
                  "value": 5,
                },
                Object {
                  "name": "losses",
                  "value": 20,
                },
                Object {
                  "name": "wins",
                  "value": 48,
                },
                Object {
                  "name": "draws",
                  "value": 23,
                },
                Object {
                  "name": "fwd_pass",
                  "value": 1533,
                },
                Object {
                  "name": "goal_assist",
                  "value": 2,
                },
                Object {
                  "name": "appearances",
                  "value": 80,
                },
                Object {
                  "name": "mins_played",
                  "value": 6953,
                },
                Object {
                  "name": "backward_pass",
                  "value": 308,
                },
              ],
            },
            Object {
              "player": Object {
                "age": "33 years 67 days",
                "currentTeam": Object {
                  "id": 11,
                  "name": "Manchester City",
                  "shortName": "Man City",
                  "teamType": "FIRST",
                },
                "id": 4148,
                "info": Object {
                  "position": "M",
                  "positionInfo": "Centre Defensive Midfielder",
                  "shirtNum": 42,
                },
                "name": Object {
                  "first": "Yaya",
                  "last": "TourÃ©",
                },
                "nationalTeam": Object {
                  "country": "Cote D'Ivoire",
                  "demonym": "Ivorian",
                  "isoCode": "CI",
                },
              },
              "stats": Array [
                Object {
                  "name": "goals",
                  "value": 65,
                },
                Object {
                  "name": "losses",
                  "value": 49,
                },
                Object {
                  "name": "wins",
                  "value": 149,
                },
                Object {
                  "name": "draws",
                  "value": 35,
                },
                Object {
                  "name": "fwd_pass",
                  "value": 4491,
                },
                Object {
                  "name": "goal_assist",
                  "value": 35,
                },
                Object {
                  "name": "appearances",
                  "value": 232,
                },
                Object {
                  "name": "mins_played",
                  "value": 18919,
                },
                Object {
                  "name": "backward_pass",
                  "value": 1995,
                },
              ],
            },
            Object {
              "player": Object {
                "age": "30 years 269 days",
                "currentTeam": Object {
                  "id": 12,
                  "name": "Manchester United",
                  "shortName": "Man Utd",
                  "teamType": "FIRST",
                },
                "id": 2064,
                "info": Object {
                  "position": "F",
                  "positionInfo": "Left/Centre/Right Second Striker",
                  "shirtNum": 10,
                },
                "name": Object {
                  "first": "Wayne",
                  "last": "Rooney",
                },
                "nationalTeam": Object {
                  "country": "England",
                  "demonym": "English",
                  "isoCode": "GB-ENG",
                },
              },
              "stats": Array [
                Object {
                  "name": "goals",
                  "value": 201,
                },
                Object {
                  "name": "losses",
                  "value": 91,
                },
                Object {
                  "name": "wins",
                  "value": 280,
                },
                Object {
                  "name": "draws",
                  "value": 90,
                },
                Object {
                  "name": "fwd_pass",
                  "value": 1795,
                },
                Object {
                  "name": "goal_assist",
                  "value": 84,
                },
                Object {
                  "name": "appearances",
                  "value": 461,
                },
                Object {
                  "name": "mins_played",
                  "value": 27056,
                },
                Object {
                  "name": "backward_pass",
                  "value": 1928,
                },
              ],
            },
            Object {
              "player": Object {
                "age": "31 years 294 days",
                "currentTeam": Object {
                  "id": 1,
                  "name": "Arsenal",
                  "shortName": "Arsenal",
                  "teamType": "FIRST",
                },
                "id": 4246,
                "info": Object {
                  "position": "D",
                  "positionInfo": "Centre Central Defender",
                  "shirtNum": 4,
                },
                "name": Object {
                  "first": "Per",
                  "last": "Mertesacker",
                },
                "nationalTeam": Object {
                  "country": "Germany",
                  "demonym": "German",
                  "isoCode": "DE",
                },
              },
              "stats": Array [
                Object {
                  "name": "goals",
                  "value": 8,
                },
                Object {
                  "name": "losses",
                  "value": 45,
                },
                Object {
                  "name": "wins",
                  "value": 117,
                },
                Object {
                  "name": "draws",
                  "value": 41,
                },
                Object {
                  "name": "fwd_pass",
                  "value": 4257,
                },
                Object {
                  "name": "appearances",
                  "value": 187,
                },
                Object {
                  "name": "mins_played",
                  "value": 16531,
                },
                Object {
                  "name": "backward_pass",
                  "value": 535,
                },
              ],
            },
            Object {
              "player": Object {
                "age": "25 years 149 days",
                "currentTeam": Object {
                  "id": 26,
                  "name": "Leicester City",
                  "shortName": "Leicester",
                  "teamType": "FIRST",
                },
                "id": 8983,
                "info": Object {
                  "position": "M",
                  "positionInfo": "Left/Right Winger",
                  "shirtNum": 26,
                },
                "name": Object {
                  "first": "Riyad",
                  "last": "Mahrez",
                },
                "nationalTeam": Object {
                  "country": "Algeria",
                  "demonym": "Algerian",
                  "isoCode": "DZ",
                },
              },
              "stats": Array [
                Object {
                  "name": "goals",
                  "value": 22,
                },
                Object {
                  "name": "losses",
                  "value": 23,
                },
                Object {
                  "name": "wins",
                  "value": 35,
                },
                Object {
                  "name": "draws",
                  "value": 21,
                },
                Object {
                  "name": "fwd_pass",
                  "value": 687,
                },
                Object {
                  "name": "goal_assist",
                  "value": 14,
                },
                Object {
                  "name": "appearances",
                  "value": 71,
                },
                Object {
                  "name": "mins_played",
                  "value": 5368,
                },
                Object {
                  "name": "backward_pass",
                  "value": 323,
                },
              ],
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
